File: internal/server/ui_dist/js/util/pricing.js
// Token-cost estimation for the Stats page. Rates are published by the
// backend (/api/metrics/stats "pricing" key or GET /api/metrics/pricing) in
// USD per million tokens; this module resolves the effective rates for a
// model (model config > UI defaults > server defaults) and converts the
// result to the display currency.
import { formatMoney } from "./format.js";

/** First non-null numeric value among the arguments, else fallback. */
function pick(...values) {
  for (const v of values) {
    if (v != null && Number.isFinite(v)) return v;
  }
  return null;
}

/**
 * Resolve the effective USD-per-million-token rates for a model.
 *
 * Precedence: a model's own pricing block is authoritative (an omitted
 * cached rate bills cached tokens at the model's input rate). Models
 * without one use the UI default rates per field, falling through to the
 * server's pricing.defaults.
 *
 * @param {string} modelName Real model ID as stored in the activity log.
 * @param {object|null} server Server pricing snapshot: {currency, usd_to_inr,
 *   defaults, models: {modelId: {input, output, cached}}}.
 * @param {object|null} uiDefaults UI default-rates override; null fields
 *   fall through to the server defaults.
 */
export function resolveRates(modelName, server, uiDefaults) {
  const modelRates = server?.models?.[modelName];
  if (modelRates) {
    return {
      input: modelRates.input ?? 0,
      output: modelRates.output ?? 0,
      // null cached bills cached tokens at the input rate
      cached: modelRates.cached ?? null,
    };
  }
  const serverDefaults = server?.defaults ?? {};
  const src = uiDefaults ?? {};
  return {
    input: pick(src.input, serverDefaults.input, 0),
    output: pick(src.output, serverDefaults.output, 0),
    // null cached bills cached tokens at the input rate
    cached: pick(src.cached, serverDefaults.cached),
  };
}

/**
 * Estimate a model's cost in USD from a stats row.
 * Cached tokens are a subset of the prompt tokens, so they are deducted from
 * the billable input and charged at the cached rate (input rate when unset).
 * Returns 0 when no rates are configured.
 *
 * @param {{input_tokens: number, output_tokens: number, cached_tokens: number}} row
 * @param {{input: number, output: number, cached: number|null}} rates
 */
export function estimateCostUSD(row, rates) {
  const cachedRate = rates.cached == null ? rates.input : rates.cached;
  const billableInput = Math.max(0, (row.input_tokens || 0) - (row.cached_tokens || 0));
  const cost =
    (billableInput * rates.input +
      (row.cached_tokens || 0) * cachedRate +
      (row.output_tokens || 0) * rates.output) /
    1e6;
  return cost;
}

/** USD -> display currency conversion (rates are dollar-denominated). */
export function toDisplayCurrency(usd, currency, usdToInr) {
  return currency === "INR" ? usd * usdToInr : usd;
}

/** Currency symbol for the effective display currency. */
export function currencySymbol(currency) {
  return currency === "INR" ? "₹" : "$";
}

/**
 * Format an estimated cost for display, applying the currency conversion and
 * adaptive precision. Returns "—" when nothing is billable or configured.
 */
export function formatCost(usd, currency, usdToInr) {
  return formatMoney(toDisplayCurrency(usd, currency, usdToInr), currencySymbol(currency));
}

File: internal/config/pricing.go
package config

import (
	"fmt"
	"strings"
)

// DefaultUSDToINR is the USD -> INR conversion rate used when
// pricing.usdToINR is not configured.
const DefaultUSDToINR = 95.0

// PricingRates holds approximate token prices in US dollars per million
// tokens. Cached is optional: a nil value bills cached tokens at the input
// rate.
type PricingRates struct {
	Input  float64  `yaml:"input"`
	Output float64  `yaml:"output"`
	Cached *float64 `yaml:"cached"`
}

// PricingConfig is the top-level `pricing:` configuration section. It defines
// the display currency, the USD -> INR conversion rate, and default token
// prices applied to models that do not define their own pricing.
type PricingConfig struct {
	// Currency selects the display currency: USD (default) or INR.
	Currency string `yaml:"currency"`

	// USDToINR converts dollar-denominated prices when Currency is INR.
	USDToINR float64 `yaml:"usdToINR"`

	// Defaults apply to models without their own pricing block.
	Defaults PricingRates `yaml:"defaults"`
}

func (r PricingRates) Validate(what string) error {
	if r.Input < 0 {
		return fmt.Errorf("%s.input: must be >= 0, got %g", what, r.Input)
	}
	if r.Output < 0 {
		return fmt.Errorf("%s.output: must be >= 0, got %g", what, r.Output)
	}
	if r.Cached != nil && *r.Cached < 0 {
		return fmt.Errorf("%s.cached: must be >= 0, got %g", what, *r.Cached)
	}
	return nil
}

// Validate normalizes and validates the pricing configuration.
func (p *PricingConfig) Validate() error {
	p.Currency = strings.ToUpper(strings.TrimSpace(p.Currency))
	switch p.Currency {
	case "":
		p.Currency = "USD"
	case "USD", "INR":
	default:
		return fmt.Errorf("pricing.currency: invalid currency %q (valid: USD, INR)", p.Currency)
	}
	if p.USDToINR == 0 {
		p.USDToINR = DefaultUSDToINR
	}
	if p.USDToINR < 0 {
		return fmt.Errorf("pricing.usdToINR: must be > 0, got %g", p.USDToINR)
	}
	return p.Defaults.Validate("pricing.defaults")
}
